### PR TITLE
Add trash

### PR DESCRIPTION
**Pull request by Athozus**
This is a PR for adding trash system to mail, as requested by some players across servers (and myself too). This would complete #38 if merged.

- [x] Add trash tab on main UI
- [x] Permit trash a mail
- [x] Enable or disable trashing via settings (can be still permanent delete)
- [x] Permit restore mails in there old boxes

Some interesting infos :
- It adds a `previous_boxes` value to mail while they're in trash (removed after restoring)
- To obtain the previous one, a new function called `mail.get_message_boxes(playername, msg_id)` returns a list of boxes like `{"inbox", "outbox"}`
- A parameter `delete_in_trash` has been added to `mail.delete_mail()` to avoid trash delete when you just moved the mails.

I'm waiting for your answers to know if you might fix things. Actually tried, it works without crash.

Some screenshots below :

1. Some mails in my trash
2. A mail has been restored
3. Setting is enabled

screenshot_20230613_201426

screenshot_20230613_201416

screenshot_20230613_202151

### CODE REVIEW

File: storage.lua [hunk 1]
```diff
@@ -135,8 +173,53 @@ function mail.delete_mail(playername, msg_ids)
 			end
 		end
 	end
+	if delete_in_trash then
+		for i = #entry.trash, 1, -1 do
+			for _, deleted_msg in ipairs(msg_ids) do
+				if entry.trash[i].id == deleted_msg then
+					table.remove(entry.trash, i)
```
Comment (BuckarooBanzay) [inline storage.lua:180]: needs a `break` after that, otherwise it may trigger the issue we had in #97 again :smirk:

**Review by BuckarooBanzay — COMMENTED** [review]
didn't test, only had one issue, otherwise looks good i think :+1:

File: storage.lua (hunk 1 above)
Comment (Athozus) [reply storage.lua:180]: Done

**Comment by S-S-X** [comment]
Didn't test but also didn't spot any obvious issues.

Minor thing but for UI I think it would be better to hide trash tab completely if trash is disabled. Doing so would allow having cleaner main user interface for those who do not want that functionality.

**Comment by Athozus** [comment]
> Minor thing but for UI I think it would be better to hide trash tab completely if trash is disabled. Doing so would allow having cleaner main user interface for those who do not want that functionality.

Hi, I'm not sure with that one because the trash is still there when you disable auto-trashing. Maybe a Trash button, separated from the Delete one, would be better ? And if the autotrash is enabled, then Trash and Delete button do the same thing.

**Comment by Athozus** [comment]
You know what, maybe we should merge know and change minor things based on user experience ?

**Comment by S-S-X** [comment]
Could be another button, but that suggestion is just based on my personal preference: to not see those buttons/tabs at all because there's a lot of stuff anyway and I would not need those for anything. From this view point adding another button would be even worse.

It is debatable if hiding existing message is bad or good thing, some could actually like that feature. But anyway this isn't really useful reasoning for doing so (but also not against doing so), just a note that in the end hiding existing(mails in trash storage) things might actually be usable feature.

**Comment by Athozus** [comment]
Following your suggestions, I merge

**Pull request merged**